resources/lib/UnitySQL.php

```diff
@@ -367,8 +383,482 @@ public function roleAvailableInGroup($uid, $group, $role)
+
+        $stmt->execute();
+
+        $roles = array();
+        foreach ($stmt->fetchAll() as $row) {
+            $roles[] = $row['role'];
+        }
+
+        return $roles;
+    }
+
+    public function getDefaultRole($group_type)
+    {
+        $stmt = $this->conn->prepare(
+            "SELECT * FROM " . self::TABLE_GROUP_TYPES . " WHERE slug=:slug"
+        );
+        $stmt->bindParam(":slug", $group_type);
+
+        $stmt->execute();
+
+        $row = $stmt->fetchAll()[0];
+        return $row['def_role'];
+    }
+
+    public function getGroupTypeDetails($group_type)
+    {
+        $stmt = $this->conn->prepare(
+            "SELECT * FROM " . self::TABLE_GROUP_TYPES . " WHERE slug=:slug"
+        );
+        $stmt->bindParam(":slug", $group_type);
+
+        $stmt->execute();
+
+        $row = $stmt->fetchAll()[0];
+
+        $out = array();
+        $out['name'] = $row['name'];
+        $out['color'] = $row['color'];
+
+        $av_roles = explode(",", $row['av_roles']);
+        $out['av_roles'] = $av_roles;
+
+        return $out;
+    }
+
+    public function getRoleName($role)
+    {
+        $stmt = $this->conn->prepare(
+            "SELECT * FROM " . self::TABLE_GROUP_ROLES . " WHERE slug=:slug"
+        );
+        $stmt->bindParam(":slug", $role);
+
+        $stmt->execute();
+
+        $row = $stmt->fetchAll()[0];
+        return $row['name'];
+    }
+
+    public function getPermissions($roles)
+    {
+        $stmt = $this->conn->prepare(
+            "SELECT * FROM " . self::TABLE_GROUP_ROLES . " WHERE slug=:slug"
+        );
+
+        $perms = array();
+        foreach ($roles as $role) {
+            $stmt->bindParam(":slug", $role);
+
+            $stmt->execute();
+
+            $row = $stmt->fetchAll()[0];
+            $perms = array_merge($perms, explode(",", $row['perms']));
+        }
+
+        return $perms;
+    }
+
+    public function getUsersWithRoles($role, $group_uid)
+    {
+        $stmt = $this->conn->prepare(
+            "SELECT * FROM " . self::TABLE_GROUP_ROLE_ASSIGNMENTS . " WHERE `group`=:group AND role=:role"
+        );
+
+        $stmt->bindParam(":group", $group_uid);
+        $stmt->bindParam(":role", $role);
+
+        $stmt->execute();
+
+        $users = array();
+        foreach ($stmt->fetchAll() as $row) {
+            $users[] = $row['user'];
+        }
+
+        return $users;
+    }
+
+    public function getUsersWithoutRoles($group_uid, $curr_users_uids)
+    {
+        $stmt = $this->conn->prepare(
+            "SELECT * FROM " . self::TABLE_GROUP_ROLE_ASSIGNMENTS . " WHERE `group`=:group"
+        );
+
+        $stmt->bindParam(":group", $group_uid);
+
+        $stmt->execute();
+
+        $users = array();
+        foreach ($stmt->fetchAll() as $row) {
+            $users[] = $row['user'];
+        }
+
+        $users = array_diff($curr_users_uids, $users);
+
+        return $users;
+    }
+
+    public function assignRole($role, $uid, $gid)
```

Comment:
Getting a fatal error when trying to approve a group request for a class/center using the user `user1@org1.test`. I tried approving the class group using the same user and got this:

```
Notice: Trying to access array offset on value of type bool in /var/www/unity-web-portal/resources/lib/UnitySQL.php on line 646

Fatal error: Uncaught PDOException: SQLSTATE[23000]: Integrity constraint violation: 1048 Column 'user' cannot be null in /var/www/unity-web-portal/resources/lib/UnitySQL.php:524 Stack trace: #0 /var/www/unity-web-portal/resources/lib/UnitySQL.php(524): PDOStatement->execute() #1 /var/www/unity-web-portal/resources/lib/UnitySQL.php(751): UnityWebPortal\lib\UnitySQL->assignRole() #2 /var/www/unity-web-portal/resources/lib/UnityGroup.php(538): UnityWebPortal\lib\UnitySQL->assignSuperRole() #3 /var/www/unity-web-portal/resources/lib/UnityGroup.php(172): UnityWebPortal\lib\UnityGroup->init() #4 /var/www/unity-web-portal/webroot/admin/group-mgmt.php(25): UnityWebPortal\lib\UnityGroup->approveGroup() #5 {main} thrown in /var/www/unity-web-portal/resources/lib/UnitySQL.php on line 524
```

I haven't tracked down the origin of this